File: Stack/132.java
class Solution {
    public boolean find132pattern(int[] nums) {

        Stack<Integer> st = new Stack<>();
        int th = Integer.MIN_VALUE;

        for (int i = nums.length - 1; i >= 0; i--) {
            if (nums[i] < th)
                return true;
            while (!st.isEmpty() && st.peek() < nums[i]) {
                th = st.pop();
            }
            st.push(nums[i]);
        }
        return false;
    }

    public boolean find132pattern(int[] nums) {
        int n = nums.length, top = n, third = Integer.MIN_VALUE;

        for (int i = n - 1; i >= 0; i--) {
            if (nums[i] < third)
                return true;
            while (top < n && nums[i] > nums[top])
                third = nums[top++];
            nums[--top] = nums[i];
        }

        return false;
    }
}

// https://leetcode.com/problems/132-pattern/

File: Stack/132.go
package stack

type Pair struct {
	value int
	min   int
}

func find132pattern(nums []int) bool {
	var stack []Pair
	currMin := nums[0]

	for i := 1; i < len(nums); i++ {
		for len(stack) > 0 && nums[i] >= stack[len(stack)-1].value {
			stack = stack[:len(stack)-1]
		}

		if len(stack) > 0 && nums[i] > stack[len(stack)-1].min {
			return true
		}
		stack = append(stack, Pair{value: nums[i], min: currMin})
		currMin = min(nums[i], currMin)
	}
	return false
}
